### PR TITLE
[JAX] Bugfix for softmax primitives accepting invalid input sharding

### PR DESCRIPTION
**Pull request by denera**
With this PR, all softmax primitives now issue a warning whenever the input hidden dimension is sharded (unsupported) and force XLA to unshard it.

The distributed unit tests are also updated to test softmax with intentionally invalid sharding to make sure it issues the warning message and returns the correct numerical result.

### CODE REVIEW

**Comment by nouiz** [comment]
cudnn_frontend is changed. Is that normal?

**Comment by denera** [comment]
> cudnn_frontend is changed. Is that normal?

No, this is not intentional. The cudnn-frontend version was updated between when I created my local branch and when I submitted the PR, but my branch did not pick that up when I rebased. I pushed the correction now and the change list no longer shows anything for cudnn-frontend.

File: transformer_engine/jax/cpp_extensions.py [hunk 1]
```diff
@@ -1211,24 +1223,38 @@ def backward_batcher(primitive, batched_args, batch_dims, *, scale_factor):
...
+        dz_spec = get_padded_spec(arg_infos[0])
+        softmax_out_spec = get_padded_spec(arg_infos[1])
+        if dz_spec[-1] is not None or softmax_out_spec[-1] is not None:
+            warnings.warn(
+                f"Sharding the hidden dimension is not supported in {cls.name}! " \
+                f"Forcing XLA to not shard the hidden dim, which might introduce extra " \
+                f"collective ops and hurt performance."
+            )
+        dz_spec = NamedSharding(mesh, PartitionSpec(*dz_spec[:-1], None))
+        softmax_out_spec = NamedSharding(mesh, PartitionSpec(*softmax_out_spec[:-1], None))
```
Comment (nouiz) [inline transformer_engine/jax/cpp_extensions.py:1257]: Should we foce dz_spec and softmax_out_spec sharding to be the same?

File: transformer_engine/jax/cpp_extensions.py [hunk 2]
```diff
@@ -1296,13 +1322,15 @@ def batcher(batched_args, batch_dims, *, scale_factor):
 
     @staticmethod
     def infer_sharding_from_operands(scale_factor, mesh, arg_infos, result_infos):
-        return SoftmaxPrimitive.forward_infer_sharding_from_operands(scale_factor, mesh, arg_infos,
```
Comment (nouiz) [inline transformer_engine/jax/cpp_extensions.py:1299]: Why this change and the one below?
That seem to do nothing, can you confirm?

File: tests/jax/test_distributed_softmax.py [hunk 3]
```diff
@@ -35,11 +37,32 @@ def generate_inputs(self, shape, mesh_resource, softmax_type, dtype):
...
             mask = make_self_mask(batch, sqelen)
 
-        x_pspec = PartitionSpec(mesh_resource.dp_resource, mesh_resource.tp_resource, None, None)
+        if not bad_sharding:
+            x_pspec = PartitionSpec(mesh_resource.dp_resource, mesh_resource.tp_resource,
+                                    None, None)
+        else:
+            x_pspec = PartitionSpec(mesh_resource.dp_resource, None,
+                                    None, mesh_resource.tp_resource)
         mask_pspec = PartitionSpec(mesh_resource.dp_resource, None, None, None)
```
Comment (nouiz) [inline tests/jax/test_distributed_softmax.py:46]: Is it useful to test bad sharding for the mask too?

**Review by nouiz — COMMENTED** [review]
A few small comment, but otherwise look good.
Ask Ming to also review.

**Review by nouiz — APPROVED** [review]
LGTM. If Ming or Reese can review, that would be great.

File: tests/jax/test_distributed_softmax.py [hunk 4]
```diff
@@ -81,3 +95,51 @@ def ref_func(x, mask):
                         metric_bwd_dtype=dtype,
                         in_shardings=(x_pspec, mask_pspec),
                         out_shardings=(None, (x_pspec,)))
+
+    @pytest.mark.parametrize('device_count,mesh_shape,mesh_axes,mesh_resource', generate_configs())
+    @pytest.mark.parametrize('data_shape', [[32, 12, 128, 128]])
+    @pytest.mark.parametrize(
+        'softmax_type',
+        [SoftmaxType.SCALED, SoftmaxType.SCALED_MASKED, SoftmaxType.SCALED_UPPER_TRIANG_MASKED])
+    def test_softmax_bad_sharding(self, device_count, mesh_shape, mesh_axes, mesh_resource, data_shape,
```
Comment (zlsh80826) [inline tests/jax/test_distributed_softmax.py:104]: Can't we just reuse the `test_softmax_sharding` with `@pytest.mark.parametrize('bad_sharding', [True, False])` and only check the warning messages when `bad_sharding == True`?

**Comment by zlsh80826** [comment]
I have an open question that is not intended to be answered in this PR. Do we need three different custom calls for the softmax? I think a single custom call with something like SoftmaxType would be sufficient to handle these three different softmax variations.

File: tests/jax/test_distributed_softmax.py [hunk 5]
```diff
@@ -81,3 +95,51 @@ def ref_func(x, mask):
...
+                    # when XLA is forced to unshard the hidden dimension. We can catch
+                    # and ignore that specific error here.
+                    if "Expected collective count" not in str(e):
+                        raise e
+                    else:
+                        for warn in w:
+                            assert "Sharding the hidden dimension is not supported" in str(warn), (
+                                "Softmax primitive did not raise the correct warning for "
+                                "unsupported sharding in the hidden dimension."
+                            )
```
[the remark this answers is not in the document]
Comment (denera) [reply tests/jax/test_distributed_softmax.py:141]: Yes, you're right. I misunderstood how `warnings.catch_warnings()` was meant to work. Thanks!

[the remark this answers is not in the document]
Comment (zlsh80826) [reply tests/jax/test_distributed_softmax.py:141]: ok, I will approve after this fix, thanks.

**Comment by zlsh80826** [comment]
LGTM! Thanks for the fixes.
